README.md
# exemplos-book

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar exemplos-book-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2020 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-12-14
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-12-14
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/exemplos-book/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/exemplos-book/compare/0.1.0...0.1.1


test/exemplos_book/core_test.clj
(ns exemplos-book.core-test
  (:require [clojure.test :refer :all]
            [exemplos-book.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/exemplos_book/core.clj
(ns exemplos-book.core
  (:require [exemplos-book.financeiro :as financeiro])
  (:gen-class))

(def transacao {:valor 200 :tipo "receita"})
(def transacao-desnecessaria {:valor 34
                              :tipo "despesa"
                              :rotulos '("desnecessaria" "cartao")})
(def transacoes
  [{:valor 33.0
    :tipo "despesa"
    :comentario "Almoço"
    :moeda "R$"
    :data "19/11/2016"}
   {:valor 2700.0
    :tipo "receita"
    :comentario "Bico"
    :moeda "R$"
    :data "01/12/2016"}
   {:valor 123.0
    :tipo "despesa"
    :comentario "Livro de Clojure"
    :moeda "R$"
    :data "03/12/2016"}])

(filter financeiro/valor-grande? transacoes)

;(:rotulos transacao-desnecessaria)
;(:tipo transacao)

(defn valor-sinalizado [transacao]
  (if (= (:tipo transacao) "despesa")
    (str "-" (:valor transacao))
    (str "+" (:valor transacao))))


src/exemplos_book/pureza.clj
(ns exemplos-book.pureza
  (:require [clojure.string :as s]))

(def de-para [{:de "a" :para "4"}
              {:de "e" :para "3"}
              {:de "i" :para "1"}
              {:de "o" :para "0"}])

(defn escrita-hacker [texto dicionario]
  (if (empty? dicionario)
    texto
    (let [conversao (first dicionario)]
      (escrita-hacker (s/replace texto
                                 (:de conversao)
                                 (:para conversao))
                      (rest dicionario)))))

(escrita-hacker "alameda" de-para)

(def cotacoes
  {:yuan {:cotacao 2.15M :simbolo "¥"}
   :euro {:cotacao 0.28M :simbolo "€"}})

;;impuro
(defn transacao-em-outra-moeda-inpuro [moeda transacao]
  (let [{{cotacao :contacao simbolo :simbolo} moeda} cotacoes]
    (assoc transacao :valor (* cotacao (:valor transacao))
           :moeda simbolo)))

;; puro
(defn transacao-em-outra-moeda-pura [cotacoes moeda transacao]
  (let [{{cotacao :contacao simbolo :simbolo} moeda} cotacoes]
    (assoc transacao :valor (* cotacao (:valor transacao))
           :moeda simbolo)))


src/exemplos_book/lazyness.clj
(ns exemplos-book.lazyness)

(defn test-da-preguica []
  (prn "não deveria aparecer nada aqui") "nada")

(defn um-oi [a b]
  "oi")

(def transacoes
  [{:valor 33.0M :tipo "despesa" :comentario "Almoço" :moeda "R$" :data "19/11/2016"}
   {:valor 2700.0M :tipo "receita" :comentario "Bico" :moeda "R$" :data "01/12/2016"}
   {:valor 29.0M :tipo "despesa" :comentario "Livro de Clojure" :moeda "R$" :data "03/12/2016"}])

(um-oi (test-da-preguica) (test-da-preguica))

;; rand-nth retorna um valor aleatório dentro de uma coleção
(rand-nth ["despesa" "receita"])
;; "despesa"
;; seu resultado pode ser diferente

;; rand-int retorna um número inteiro aleatório entre 0 e
;; o argumento, sem incluir o argumento nas possibilidades
;; daí multiplicamos por 0.01M para ter um número real com
;; duas casas decimais
(* (rand-int 100001) 0.01M)
;; 243.85M
;; seu resultado pode ser diferente

(defn transacao-aleatoria []
  {:valor (* (rand-int 100001) 0.01M)
   :tipo (rand-nth ["despesa" "receita"])})

(transacao-aleatoria)

(repeatedly 3 transacao-aleatoria)

(class (repeatedly 3 transacao-aleatoria))

(def transacoes-aleatorias (repeatedly transacao-aleatoria))

(take 1 transacoes-aleatorias)
(take 3 transacoes-aleatorias)
(take 5 transacoes-aleatorias)
(= (take 5 transacoes-aleatorias) (take 5 transacoes-aleatorias))
(take 50 transacoes-aleatorias)

(cons (transacao-aleatoria) transacoes)

(defn aleatorias2
;; começando a sequência criando a primeira transação
  ([quantidade]
   (aleatorias2 quantidade 1 (list (transacao-aleatoria))))
;; incrementando a sequência até que a quantidade desejada
;; seja atendida
  ([quantidade quantas-ja-foram transacoes]
   (if (< quantas-ja-foram quantidade)
     (aleatorias2 quantidade (inc quantas-ja-foram)
                  (cons (transacao-aleatoria)
                        transacoes))
     transacoes)))

(aleatorias2 4)
(aleatorias2 900000)

(defn aleatorias3
  ([quantidade]
   (aleatorias3 quantidade 1 (list (transacao-aleatoria))))
  ([quantidade quantas-ja-foram transacoes]
   (if (= quantas-ja-foram quantidade)
     transacoes
;; aplicando a otimização na cauda
     (recur quantidade (inc quantas-ja-foram)
            (cons (transacao-aleatoria) transacoes)))))

(defn aleatorias
  ([quantidade]
   (aleatorias quantidade 1 (list (transacao-aleatoria))))
  ([quantidade quantas-ja-foram transacoes]
 ;; vamos embalar a parte recursiva na macro lazy-seq
   (lazy-seq
    (if (= quantas-ja-foram quantidade)
      transacoes
      (aleatorias quantidade (inc quantas-ja-foram)
                  (cons (transacao-aleatoria)
                        transacoes))))))

(aleatorias 4)
(aleatorias 900000)

(class (aleatorias 4))
(class (aleatorias 900000))

(time (class (aleatorias 4)))
(time (class (aleatorias 900000)))

(defn aleatorias-infi []
  (lazy-seq
   (cons (transacao-aleatoria) (aleatorias-infi))))

(time (class (take 4 (aleatorias-infi))))
(time (class (take 900000 (aleatorias-infi))))

(take 4 (aleatorias-infi))
(take 900000 (aleatorias-infi))


src/exemplos_book/financeiro.clj
(ns exemplos-book.financeiro
  (:require [clojure.string :as s]))

(def transacoes
  [{:valor 33.0 :tipo "despesa" :comentario "Almoço" :data "19/11/2016"}
   {:valor 2700.0 :tipo "receita" :comentario "Bico" :data "01/12/2016"}
   {:valor 29.0 :tipo "despesa" :comentario "Livro de Clojure" :data "03/12/2016"}];; verifica se a transacao é do tipo despesa
(defn despesa? [transacao]
  (= (:tipo transacao) "despesa"))

;; cria resumo de uma transacao
(defn resumo [transacao]
  (select-keys transacao [:valor :tipo :data]))

;; pega só o valor das transacoes
(defn soma-valor [transacao]
  (:valor transacao))

;; verifica se o valor da transacao é maior que 100
(defn valor-grande? [transacao]
  (> (:valor transacao) 100))

;; filtra valores acima de 100
;;(filter valor-grande? transacoes)

;; filtra valores acima de 100 sem a funcao valor-grande?
(filter (fn [transacao]
          (> (:valor transacao) 100))
        transacoes)
;;ou
(filter #(> (:valor %) 100) transacoes)

;; filtra um conjunto de valores de acordo com uma função
(filter despesa? transacoes)

;; mapea um conjunto de valores de acordo com uma função
(map resumo transacoes)

;; pega só os valores das despesas
(map soma-valor (filter despesa? transacoes))

;; soma todos os valores das despesas
(reduce +
        (map soma-valor
             (filter despesa?
                     transacoes)))
;; resumo disso acima
(reduce + (map #(:valor %)
               (filter #(= (:tipo %) "despesa")
                       transacoes)))

(soma-valor (first transacoes))

(-> (first transacoes)
    (soma-valor))

(def transacoes
  [{:valor 33.0M
    :tipo "despesa"
    :comentario "Almoço"
    :moeda "R$"
    :data "19/11/2016"}
   {:valor 2700.0M
    :tipo "receita"
    :comentario "Bico"
    :moeda "R$"
    :data "01/12/2016"}
   {:valor 29.0M
    :tipo "despesa"
    :comentario "Livro de Clojure"
    :moeda "R$"
    :data "03/12/2016"}])

(comment
  (defn valor-sinalizado1 [transacao]
    (if (= (:tipo transacao) " despesa ")
      (str (:moeda transacao) " - " (:valor transacao))
      (str (:moeda transacao) " + " (:valor transacao)))))

(defn valor-sinalizado [transacao]
  (let [moeda (:moeda transacao "R$")
        valor (:valor transacao)]
    (if (= (:tipo transacao) "despesa")
      (str moeda " -" valor)
      (str moeda " +" valor))))

(valor-sinalizado (first transacoes))
(valor-sinalizado (second transacoes))

(def transacao-aleatoria {:valor 9.0})

(valor-sinalizado transacao-aleatoria)

(defn data-valor [transacao]
  (str (:data transacao) " => " (valor-sinalizado transacao)))

(data-valor (first transacoes))

(def cotacoes
  {:yuan {:cotacao 2.15M :simbolo "¥"}
   :euro {:cotacao 0.28M :simbolo "€"}})

;;(defn transacao-em-yuan [transacao]
;;  (assoc transacao :valor (* (:cotacao (:yuan cotacoes))
;;                             (:valor transacao))
;;        :moeda (:simbolo (:yuan cotacoes))))

;(defn transacao-em-yuan [transacao]
;  (assoc transacao :valor (* (get-in cotacoes [:yuan :cotacao])
;                             (:valor transacao))
;         :moeda (get-in cotacoes [:yuan :simbolo])))

(defn transacao-em-yuan [transacao]
  (let [yuan (:yuan cotacoes)]
    (assoc transacao :valor (* (:cotacao yuan (:valor transacao))
                               (:valor transacao))
           :moeda (:simbolo yuan))))

(data-valor (first transacoes))
(data-valor (transacao-em-yuan (first transacoes)))

(transacao-em-yuan (first transacoes))

(defn texto-resumo-em-yuan [transacao]
  (-> (transacao-em-yuan transacao)
      (data-valor)))

;;(map texto-resumo-em-yuan transacoes)

;;(def texto-resumo-em-yuan (comp data-valor transacao-em-yuan))

;;(map texto-resumo-em-yuan transacoes)

(defn transacao-em-yuan2 [transacao]
  (let [{{cotacao :cotacao simbolo :simbolo} :yuan} cotacoes]
    (assoc transacao :valor (* cotacao (:valor transacao))
           :moeda simbolo)))

(transacao-em-yuan2 (first transacoes))

(defn transacao-em-outra-moeda [moeda transacao]
  (let [{{cotacao :cotacao simbolo :simbolo} moeda} cotacoes]
    (assoc transacao :valor (* cotacao (:valor transacao))
           :moeda simbolo)))

(transacao-em-outra-moeda :euro (first transacoes))
(transacao-em-outra-moeda :yuan (first transacoes))


src/exemplos_book/sysFinanceiro.clj
(ns exemplos-book.sysFinanceiro
  (:gen-class)
  (:require [clojure.string :as s]))

;;data

(def transacoes
  [{:valor 33.0M :tipo "despesa" :comentario "Almoço" :moeda "R$" :data "19/11/2016"}
   {:valor 2700.0M :tipo "receita" :comentario "Bico" :moeda "R$" :data "01/12/2016"}
   {:valor 29.0M :tipo "despesa" :comentario "Livro de Clojure" :moeda "R$" :data "03/12/2016"}])

(def cotacoes
  {:yuan {:cotacao 2.15M :simbolo "¥"}
   :euro {:cotacao 0.28M :simbolo "€"}})

;;definitions
(defn valor-sinalizado [transacao]
  (let [moeda (:moeda transacao "R$")
        valor (:valor transacao)]
    (if (= (:tipo transacao) "despesa")
      (str moeda " -" valor)
      (str moeda " +" valor))))

(defn data-valor [transacao]
  (str (:data transacao) " => " (valor-sinalizado transacao)))

(defn transacao-em-yuan2 [transacao]
  (let [{{cotacao :cotacao simbolo :simbolo} :yuan} cotacoes]
    (assoc transacao :valor (* cotacao (:valor transacao))
           :moeda simbolo)))

(defn transacao-em-outra-moeda [moeda transacao]
  (let [{{cotacao :cotacao simbolo :simbolo} moeda} cotacoes]
    (assoc transacao :valor (* cotacao (:valor transacao))
           :moeda simbolo)))

;; tests

(def transacao-em-euro (partial transacao-em-outra-moeda :euro))
(def transacao-em-yuan (partial transacao-em-outra-moeda :yuan))

(defn texto-resumo-em-yuan [transacao]
  (-> (transacao-em-yuan transacao)
      (data-valor)))

(transacao-em-yuan (first transacoes))

(map transacao-em-yuan transacoes)
(map transacao-em-euro transacoes)

(s/join ", " (map texto-resumo-em-yuan transacoes))

(def juntar-tudo (partial s/join ", "))

(juntar-tudo (map texto-resumo-em-yuan transacoes))


src/exemplos_book/aridadeMultipla.clj
(ns exemplos-book.aridadeMultipla)

(def cotacoes
  {:yuan {:cotacao 2.15M :simbolo "¥"}
   :euro {:cotacao 0.28M :simbolo "€"}})

(def transacoes
  [{:valor 33.0M :tipo "despesa" :comentario "Almoço" :moeda "R$" :data "19/11/2016"}
   {:valor 2700.0M :tipo "receita" :comentario "Bico" :moeda "R$" :data "01/12/2016"}
   {:valor 29.0M :tipo "despesa" :comentario "Livro de Clojure" :moeda "R$" :data "03/12/2016"}])

(def transacoes
  [{:valor 33M :tipo "despesa" :comentario "Almoço"
    :moeda "R$" :data "19/11/2016"}
   {:valor 2700M :tipo "receita" :comentario "Bico"
    :moeda "R$" :data "01/12/2016"}
   {:valor 29M :tipo "despesa" :comentario "Livro de Clojure"
    :moeda "R$" :data "03/12/2016"}
   {:valor 45M :tipo "despesa" :comentario "Jogo no Steam"
    :moeda "R$" :data "26/12/2016"}])

(defn transacao-em-outra-moeda
  ([cotacoes moeda transacao]
   (let [{{cotacao :cotacao simbolo :simbolo} moeda} cotacoes]
     (assoc transacao :valor (* cotacao (:valor transacao))
            :moeda simbolo)))
  ([moeda transacao]
   (transacao-em-outra-moeda cotacoes  moeda transacao)))

(def membros-fundadores
  (list "Argentina" "Brasil" "Paraguai" "Uruguai"))
(def membros-plenos (cons "Venezuela" membros-fundadores))

(println membros-plenos)
(rest membros-plenos)

(identical? (rest membros-plenos)
            membros-fundadores)

(defn transacao-em-yuan [transacao]
  (let [{{cotacao :cotacao simbolo :simbolo} :yuan} cotacoes]
    (assoc transacao :valor (* cotacao (:valor transacao))
           :moeda simbolo)))

(transacao-em-yuan (first transacoes))

(def registros (atom ()))

(swap! registros conj {:valor 29M :tipo "despesa"
                       :comentario "Livro de Clojure" :moeda "R$"
                       :data "03/12/2016"})
(conj {:valor 29M :tipo "despesa" :comentario "Livro de Clojure"
       :moeda "R$" :data "03/12/2016"} ())

(swap! registros conj
       {:valor 2700M :tipo "receita" :comentario "Bico"
        :moeda "R$" :data "01/12/2016"})

(println registros)

(defn registrar [transacao]
  (swap! registros conj transacao))

(registrar {:valor 33M :tipo "despesa" :comentario "Almoço"
            :moeda "R$" :data "19/11/2016"})

(registrar {:valor 45M :tipo "despesa" :comentario "Jogo no Steam"
            :moeda "R$" :data "26/12/2016"})

(def transacoes @registros)

;; ------------ recursao
(defn despesa? [transacao]
  (= (:tipo transacao) "despesa"))

(defn saldo-acumulado1 [acumulado transacoes]
;; if-let
  (if-let [transacao (first transacoes)]
;; se _transacao_ existir, continue calculando o saldo
    (saldo-acumulado1 (if (despesa? transacao)
                        (- acumulado (:valor transacao))
                        (+ acumulado (:valor transacao)))
                      (rest transacoes))
;; se não existir, a coleção de transacoes acabou e é hora de
;; retornar o resultado
    acumulado))

(defn calcular [acumulado transacao]
  (let [valor (:valor transacao)]
    (if (despesa? transacao)
      (- acumulado valor)
      (+ acumulado valor))))

(defn saldo-acumulado2 [acumulado transacoes]
  (if-let [transacao (first transacoes)]
    (saldo-acumulado2 (calcular acumulado transacao)
                      (rest transacoes))
    acumulado))

(defn saldo-acumulado3 [acumulado transacoes]
  (if-let [transacao (first transacoes)]
    (do
      (prn "Começou saldo-acumulado. Saldo até agora:" acumulado)
      (saldo-acumulado3 (calcular acumulado transacao)
                        (rest transacoes)))
    acumulado))

(defn saldo-acumulado [acumulado transacoes]
  (if-let [transacao (first transacoes)]
    (do
      (prn "Começou saldo-acumulado. Saldo até agora:" acumulado)
      (saldo-acumulado (calcular acumulado transacao)
                       (rest transacoes)))
    (do
      (prn "Processo encerrado. Saldo final:" acumulado)
      acumulado)))

(saldo-acumulado 0 transacoes)
(saldo-acumulado 0 ())
(saldo-acumulado 0 (take 2 transacoes))

(defn saldo2 [transacoes]
  (saldo-acumulado 0 transacoes))

(saldo2 transacoes)

(defn saldo
;; caso a função receba só um argumento
  ([transacoes]
   (saldo 0 transacoes))
;; caso a função receba dois argumentos
  ([acumulado transacoes]
   (if-let [transacao (first transacoes)]
     (saldo (calcular acumulado transacao) (rest transacoes))
     acumulado)))

(saldo transacoes)

(defn como-transacao [valor]
  {:valor valor})

(def poucas-transacoes
  (map como-transacao (range 10)))

(def muitas-transacoes
  (map como-transacao (range 1000)))

(def incontaveis-transacoes
  (map como-transacao (range 100000)))

(saldo poucas-transacoes)
(saldo muitas-transacoes)
(saldo incontaveis-transacoes)

(defn saldo-tail-recursion
  ([transacoes] (saldo-tail-recursion 0 transacoes))

  ([acumulado transacoes]
   (if (empty? transacoes)
     acumulado
     (saldo-tail-recursion
      (calcular acumulado
                (first transacoes))
      (rest transacoes)))))

(saldo-tail-recursion poucas-transacoes)
(saldo-tail-recursion muitas-transacoes)
(saldo-tail-recursion incontaveis-transacoes)

;; nova versão de saldo, desta vez com recur
(defn saldo-tail-recursion-elemination
  ([transacoes] (saldo-tail-recursion-elemination 0 transacoes))
  ([acumulado transacoes]
   (if (empty? transacoes)
     acumulado
;;    vvvv aqui utilizamos recur em vez de saldo
     (recur (calcular acumulado (first transacoes))
            (rest transacoes)))))

(saldo-tail-recursion-elemination poucas-transacoes)
(saldo-tail-recursion-elemination muitas-transacoes)
(saldo-tail-recursion-elemination incontaveis-transacoes)
